File: app/src/main/java/com/clockmods/weather/QWeatherSigner.java
package com.clockmods.weather;

import net.i2p.crypto.eddsa.EdDSAPrivateKey;
import net.i2p.crypto.eddsa.EdDSASecurityProvider;
import net.i2p.crypto.eddsa.EdDSAEngine;

import java.security.PrivateKey;
import java.security.Security;
import java.security.spec.PKCS8EncodedKeySpec;

public final class QWeatherSigner {
    private static final char[] BASE64_URL =
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_".toCharArray();

    private QWeatherSigner() { }

        public static String token(String credentialId, String projectId, String privateKeyBase64,
            long nowSeconds) throws Exception {
        if (Security.getProvider(EdDSASecurityProvider.PROVIDER_NAME) == null) {
            Security.addProvider(new EdDSASecurityProvider());
        }
        String header = "{\"alg\":\"EdDSA\",\"kid\":\"" + credentialId + "\"}";
        long issuedAt = nowSeconds - 30L;
        String payload = "{\"sub\":\"" + projectId + "\",\"iat\":" + issuedAt
            + ",\"exp\":" + (issuedAt + 900L) + "}";
        String signingInput = base64Url(header.getBytes("UTF-8")) + "."
            + base64Url(payload.getBytes("UTF-8"));
        byte[] privateKeyBytes = decodeBase64(privateKeyBase64);
        PrivateKey privateKey = new EdDSAPrivateKey(new PKCS8EncodedKeySpec(privateKeyBytes));
        EdDSAEngine engine = new EdDSAEngine();
        engine.initSign(privateKey);
        byte[] signature = engine.signOneShot(signingInput.getBytes("UTF-8"));
        return signingInput + "." + base64Url(signature);
        }

    private static String base64Url(byte[] value) {
        StringBuilder encoded = new StringBuilder((value.length * 4 + 2) / 3);
        for (int index = 0; index < value.length; index += 3) {
            int first = value[index] & 0xff;
            int second = index + 1 < value.length ? value[index + 1] & 0xff : 0;
            int third = index + 2 < value.length ? value[index + 2] & 0xff : 0;
            encoded.append(BASE64_URL[first >>> 2]);
            encoded.append(BASE64_URL[((first & 0x03) << 4) | (second >>> 4)]);
            if (index + 1 < value.length) {
                encoded.append(BASE64_URL[((second & 0x0f) << 2) | (third >>> 6)]);
            }
            if (index + 2 < value.length) {
                encoded.append(BASE64_URL[third & 0x3f]);
            }
        }
        return encoded.toString();
    }

    private static byte[] decodeBase64(String value) {
        StringBuilder cleaned = new StringBuilder(value.length());
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if (!Character.isWhitespace(character) && character != '=') cleaned.append(character);
        }
        int outputLength = cleaned.length() * 6 / 8;
        byte[] decoded = new byte[outputLength];
        int buffer = 0;
        int bits = 0;
        int outputIndex = 0;
        for (int index = 0; index < cleaned.length(); index++) {
            int digit = base64Digit(cleaned.charAt(index));
            if (digit < 0) throw new IllegalArgumentException("Invalid Base64 private key");
            buffer = (buffer << 6) | digit;
            bits += 6;
            if (bits >= 8) {
                bits -= 8;
                if (outputIndex < decoded.length) decoded[outputIndex++] = (byte) (buffer >> bits);
                buffer &= (1 << bits) - 1;
            }
        }
        return decoded;
    }

    private static int base64Digit(char character) {
        if (character >= 'A' && character <= 'Z') return character - 'A';
        if (character >= 'a' && character <= 'z') return character - 'a' + 26;
        if (character >= '0' && character <= '9') return character - '0' + 52;
        if (character == '+' || character == '-') return 62;
        if (character == '/' || character == '_') return 63;
        return -1;
    }
}

File: app/src/main/java/com/clockmods/weather/QWeatherSigner.kt
package com.clockmods.weather

import net.i2p.crypto.eddsa.EdDSAPrivateKey
import net.i2p.crypto.eddsa.EdDSASecurityProvider
import net.i2p.crypto.eddsa.EdDSAEngine
import java.nio.charset.StandardCharsets
import java.security.PrivateKey
import java.security.Security
import java.security.spec.PKCS8EncodedKeySpec

object QWeatherSigner {
    private const val BASE64_URL = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"

    @JvmStatic
    @Throws(Exception::class)
    fun token(
        credentialId: String,
        developerId: String,
        projectId: String,
        privateKeyBase64: String,
        nowSeconds: Long,
    ): String {
        if (Security.getProvider(EdDSASecurityProvider.PROVIDER_NAME) == null) {
            Security.addProvider(EdDSASecurityProvider())
        }
        val header = "{\"alg\":\"EdDSA\",\"kid\":\"$credentialId\"}"
        val issuedAt = nowSeconds - 30L
        val payload = "{\"iss\":\"$developerId\",\"sub\":\"$projectId\",\"iat\":$issuedAt,\"exp\":${issuedAt + 900L}}"
        val signingInput = base64Url(header.toByteArray(StandardCharsets.UTF_8)) + "." +
            base64Url(payload.toByteArray(StandardCharsets.UTF_8))
        val privateKey: PrivateKey = EdDSAPrivateKey(PKCS8EncodedKeySpec(decodeBase64(privateKeyBase64)))
        val engine = EdDSAEngine()
        engine.initSign(privateKey)
        val signature = engine.signOneShot(signingInput.toByteArray(StandardCharsets.UTF_8))
        return "$signingInput.${base64Url(signature)}"
    }

    private fun base64Url(value: ByteArray): String {
        val encoded = StringBuilder((value.size * 4 + 2) / 3)
        var index = 0
        while (index < value.size) {
            val first = value[index].toInt() and 0xff
            val second = if (index + 1 < value.size) value[index + 1].toInt() and 0xff else 0
            val third = if (index + 2 < value.size) value[index + 2].toInt() and 0xff else 0
            encoded.append(BASE64_URL[first ushr 2])
            encoded.append(BASE64_URL[((first and 0x03) shl 4) or (second ushr 4)])
            if (index + 1 < value.size) encoded.append(BASE64_URL[((second and 0x0f) shl 2) or (third ushr 6)])
            if (index + 2 < value.size) encoded.append(BASE64_URL[third and 0x3f])
            index += 3
        }
        return encoded.toString()
    }

    private fun decodeBase64(value: String): ByteArray {
        val cleaned = buildString(value.length) {
            value.forEach { if (!it.isWhitespace() && it != '=') append(it) }
        }
        val decoded = ByteArray(cleaned.length * 6 / 8)
        var buffer = 0
        var bits = 0
        var outputIndex = 0
        cleaned.forEach { character ->
            val digit = base64Digit(character)
            require(digit >= 0) { "Invalid Base64 private key" }
            buffer = (buffer shl 6) or digit
            bits += 6
            if (bits >= 8) {
                bits -= 8
                if (outputIndex < decoded.size) decoded[outputIndex++] = (buffer shr bits).toByte()
                buffer = buffer and ((1 shl bits) - 1)
            }
        }
        return decoded
    }

    private fun base64Digit(character: Char): Int = when {
        character in 'A'..'Z' -> character - 'A'
        character in 'a'..'z' -> character - 'a' + 26
        character in '0'..'9' -> character - '0' + 52
        character == '+' || character == '-' -> 62
        character == '/' || character == '_' -> 63
        else -> -1
    }
}
